src/Rings/mpoly-ideals.jl

```diff
@@ -2515,3 +2508,35 @@ end
 function hash(I::Ideal, c::UInt)
   return hash(base_ring(I), c)
 end
+
+# Recognition of principal ideals
+function is_known(::typeof(is_principal), I::MPolyIdeal)
+  is_one(length(gens(I))) && return true
+  has_attribute(I, :is_principal) && return true
+  return false
+end
+
+@attr Bool function is_principal(I::MPolyIdeal)
+  is_one(ngens(I)) && return true
+  g = small_generating_set(I)
+  !is_one(length(g)) && return false
+  set_attribute!(I, :principal_generator=>only(g))
+  return true
+end
+
+function principal_generator(I::MPolyIdeal{T}) where {T}
+  is_one(length(gens(I))) && return only(gens(I))
+  @req is_principal(I) "ideal is not principal" # sets the attribute in particular
+  return get_attribute(I, :principal_generator)::T
+end
+ 
+ # This is a catchall method which should probably be moved elsewher
+ function is_gen(a::MPolyRingElem)
+   is_zero(a) && return false
+   !is_one(length(a)) && return false
+   e = only(AbstractAlgebra.exponent_vectors(a))
+   is_one(only(AbstractAlgebra.coefficients(a))) || return false
+   i = findfirst(is_one, e)
+   isnothing(i) && return false
+   return all(k == i || is_zero(n) for (k, n) in enumerate(e))
+ end
```

Comment:
```suggestion
 function is_gen(a::MPolyDecRingElem)
   return is_gen(forget_grading(x))
 end
```
If the issue is that this doesn't work for graded rings, then it would be better to add a method *just* for graded rings. And this should ideally delegate to the underlying ungraded ring to make use of more efficient implementations from FLINT